src/components/flowchart/draw.js

```diff
@@ -83,141 +88,165 @@ export const drawNodes = function() {
     textLabels
   } = this.props;
 
-  this.el.nodes = this.el.nodeGroup
-    .selectAll('.pipeline-node')
-    .data(nodes, node => node.id);
-
-  const enterNodes = this.el.nodes
-    .enter()
-    .append('g')
-    .attr('tabindex', '0')
-    .attr('class', 'pipeline-node');
-
-  enterNodes
-    .attr('transform', node => `translate(${node.x}, ${node.y})`)
-    .attr('opacity', 0);
-
-  enterNodes.append('rect');
-
-  enterNodes.append(icon);
-
-  enterNodes
-    .append('text')
-    .text(node => node.name)
-    .attr('text-anchor', 'middle')
-    .attr('dy', 5)
-    .attr('dx', node => node.textOffset);
-
-  this.el.nodes
-    .exit()
-    .transition('exit-nodes')
-    .duration(this.DURATION)
-    .attr('opacity', 0)
-    .remove();
-
-  this.el.nodes = this.el.nodes
-    .merge(enterNodes)
-    .attr('data-id', node => node.id)
-    .classed('pipeline-node--parameters', node => node.type === 'parameters')
-    .classed('pipeline-node--data', node => node.type === 'data')
-    .classed('pipeline-node--task', node => node.type === 'task')
-    .classed('pipeline-node--icon', !textLabels)
-    .classed('pipeline-node--text', textLabels)
-    .classed('pipeline-node--active', node => nodeActive[node.id])
-    .classed('pipeline-node--selected', node => nodeSelected[node.id])
-    .classed(
-      'pipeline-node--faded',
-      node => centralNode && !linkedNodes[node.id]
-    )
-    .on('click', this.handleNodeClick)
-    .on('mouseover', this.handleNodeMouseOver)
-    .on('mouseout', this.handleNodeMouseOut)
-    .on('focus', this.handleNodeMouseOver)
-    .on('blur', this.handleNodeMouseOut)
-    .on('keydown', this.handleNodeKeyDown);
-
-  this.el.nodes
-    .transition('update-nodes')
-    .duration(this.DURATION)
-    .attr('opacity', 1)
-    .attr('transform', node => `translate(${node.x}, ${node.y})`)
-    .end()
-    .catch(() => {})
-    .finally(() => {
-      // Sort nodes so tab focus order follows X/Y position
-      this.el.nodes.sort((a, b) => a.order - b.order);
-    });
-
-  this.el.nodes
-    .select('rect')
-    .attr('width', node => node.width - 5)
-    .attr('height', node => node.height - 5)
-    .attr('x', node => (node.width - 5) / -2)
-    .attr('y', node => (node.height - 5) / -2)
-    .attr('rx', node => (node.type === 'task' ? 0 : node.height / 2));
-
-  this.el.nodes
-    .select('.pipeline-node__icon')
-    .transition('node-icon-offset')
-    .duration(150)
-    .attr('width', node => node.iconSize)
-    .attr('height', node => node.iconSize)
-    .attr('x', node => node.iconOffset)
-    .attr('y', node => node.iconSize / -2);
+  if (changed('nodes')) {
+    this.el.nodes = this.el.nodeGroup
+      .selectAll('.pipeline-node')
+      .data(nodes, node => node.id);
+  }
+
+  if (!this.el.nodes) {
+    return;
+  }
+
+  const enterNodes = this.el.nodes.enter().append('g');
+  const exitNodes = this.el.nodes.exit();
+  const allNodes = this.el.nodes.merge(enterNodes).merge(exitNodes);
```

Comment:
Does allNodes need to include exitNodes?

Reply:
Yeah so I tried to find out if they're already included but I guessed they might not be since the enter nodes are not, so I just added that to be explicit about that either way for consistency. It's pretty subtle too but I think it's worth updating those nodes just before they exit to be sure we avoid visual glitches.

Reply:
Yeah that's what I was wondering about. Guess there's the potential for minor visual glitches, but balanced against the potential for marginal performance hits from updating those nodes even as they're fading out anyway. Guess it probably doesn't matter much either way.

Anyway, nice work, thanks for this!